src/perform_step/rkn_perform_step.jl

```diff
@@ -1819,3 +1819,94 @@ end
         integrator.EEst = integrator.opts.internalnorm(atmp, t)
     end
 end
+
+function initialize!(integrator, cache::RKN4Cache)
+    @unpack fsalfirst, k = cache
+    duprev, uprev = integrator.uprev.x
+    integrator.fsalfirst = fsalfirst
+    integrator.fsallast = k
+    integrator.kshortsize = 2
+    resize!(integrator.k, integrator.kshortsize)
+    integrator.k[1] = integrator.fsalfirst
+    integrator.k[2] = integrator.fsallast
+    integrator.f.f1(integrator.k[1].x[1], duprev, uprev, integrator.p, integrator.t)
+    integrator.f.f2(integrator.k[1].x[2], duprev, uprev, integrator.p, integrator.t)
+    integrator.stats.nf += 1
+    integrator.stats.nf2 += 1
+end
+
+@muladd function perform_step!(integrator, cache::RKN4constantCache, repeat_step = false)
+    @unpack t, dt, f, p = integrator
+    duprev, uprev = integrator.uprev.x
+    
+    #define dt values
+    halfdt = dt/2
+    dtsq = dt^2
+    eightdtsq = dtsq/8
+    halfdtsq = dtsq/2
+    sixthdtsq = dtsq/6
+    sixthdt = dt/6
+    ttmp = t + halfdt
+
+    #perform operations to find k values
+    k₁ = integrator.fsalfirst.x[1]
+    ku = uprev + halfdt * duprev + eightdtsq * k₁
+    kdu = duprev + halfdt * k₁
+
+    k₂ = f.f1(kdu, ku, p, ttmp)
+    ku = uprev + dt * duprev + halfdtsq * k₂
+    kdu = duprev + dt * k₂
+
+    k₃ = f.f1(kdu, ku, p, t + dt)
+    ku = uprev + dt * duprev + eightdtsq * k₃
+    kdu = duprev + dt * k₃
+
+    #perform final calculations to determine new y and y'.
+    u = uprev + sixthdtsq* (1*k₁ + 2*k₂ + 0*k₃) + dt * duprev
+    du = duprev + sixthdt * (1*k₁ + 4*k₂ + 1*k₃)
+
+    integrator.u = ArrayPartition((du, u))
+    integrator.fsallast = ArrayPartition((f.f1(du, u, p, t + dt), f.f2(du, u, p, t + dt)))
+    integrator.stats.nf += 3
+    integrator.stats.nf2 += 1
+    integrator.k[1] = integrator.fsalfirst
+    integrator.k[2] = integrator.fsallast
+end
+
+@muladd function perform_step!(integrator, cache::RKN4Cache, repeat_step = false)
+    @unpack t, dt, f, p = integrator
+    duprev, uprev = integrator.uprev.x
+    
+    #define dt values
+    halfdt = dt/2
+    dtsq = dt^2
+    eightdtsq = dtsq/8
+    halfdtsq = dtsq/2
+    sixthdtsq = dtsq/6
+    sixthdt = dt/6
+    ttmp = t + halfdt
+
+    #perform operations to find k values
+    k₁ = integrator.fsalfirst.x[1]
+    ku = uprev + halfdt * duprev + eightdtsq * k₁
+    kdu = duprev + halfdt * k₁
+
+    k₂ = f.f1(kdu, ku, p, ttmp)
+    ku = uprev + dt * duprev + halfdtsq * k₂
+    kdu = duprev + dt * k₂
+
+    k₃ = f.f1(kdu, ku, p, t + dt)
+    ku = uprev + dt * duprev + eightdtsq * k₃
+    kdu = duprev + dt * k₃
+
```

Comment:
This looks out of place, not in-place mutating, so it won't work.

Test the out of place one first since that looks correct.